test/e2e/tests/confirmations/navigation.spec.ts

```diff
@@ -0,0 +1,216 @@
+        await driver.switchToWindowWithTitle(WINDOW_TITLES.TestDApp);
+
+        await verifyRejectionResults(driver, '#signTypedDataResult');
+        await verifyRejectionResults(driver, '#signTypedDataV3Result');
+        await verifyRejectionResults(driver, '#signTypedDataV4Result');
+      },
+    );
+  });
+});
+
+async function verifyRejectionResults(driver: Driver, verifyResultId: string) {
+  const rejectionResult = await driver.findElement(verifyResultId);
+  assert.equal(
+    await rejectionResult.getText(),
+    'Error: User rejected the request.',
+  );
+}
+
+async function verifySignedTypeV3Confirmation(
+  driver: Driver,
+  origin: WebElement,
+) {
+  const fromAddress = driver.findElement({
+    css: '.name__value',
+    text: '0xCD2a3...DD826',
+  });
+  const toAddress = driver.findElement({
+    css: '.name__value',
+    text: '0xbBbBB...bBBbB',
+  });
+  const contents = driver.findElement({ text: 'Hello, Bob!' });
+
+  assert.ok(await origin, 'origin');
+  assert.ok(await fromAddress, 'fromAddress');
+  assert.ok(await toAddress, 'toAddress');
+  assert.ok(await contents, 'contents');
+}
+
+async function verifySignedTypeV4Confirmation(
+  driver: Driver,
+  origin: WebElement,
+) {
+  verifySignedTypeV3Confirmation(driver, origin);
+  const attachment = driver.findElement({ text: '0x' });
+  assert.ok(await attachment, 'attachment');
+}
+
+async function queueSignatures(driver: Driver) {
+  await driver.clickElement('#signTypedData');
+  await driver.waitUntilXWindowHandles(3);
+  await driver.switchToWindowWithTitle(WINDOW_TITLES.TestDApp);
+
+  await driver.clickElement('#signTypedDataV3');
+  await driver.waitUntilXWindowHandles(3);
+  await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+
+  await driver.waitUntilXWindowHandles(3);
+  await driver.switchToWindowWithTitle(WINDOW_TITLES.TestDApp);
+
+  await driver.clickElement('#signTypedDataV4');
+  await driver.waitUntilXWindowHandles(3);
+  await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+}
+
+async function queueSignaturesAndTransactions(driver: Driver) {
+  await driver.clickElement('#signTypedData');
+  await driver.waitUntilXWindowHandles(3);
+  await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+  await driver.delay(2000); // Delay deeded due to a race condition
+  // To be fixed in https://github.com/MetaMask/metamask-extension/issues/25251
+
+  await driver.waitUntilXWindowHandles(3);
+  await driver.switchToWindowWithTitle(WINDOW_TITLES.TestDApp);
+
+  await driver.clickElement('#sendButton');
+  await driver.waitUntilXWindowHandles(3);
+  await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+  await driver.delay(2000);
+
+  await driver.waitUntilXWindowHandles(3);
+  await driver.switchToWindowWithTitle(WINDOW_TITLES.TestDApp);
+
+  await driver.clickElement('#signTypedDataV3');
+  await driver.waitUntilXWindowHandles(3);
+  await driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+  await driver.delay(2000);
```

Comment:
can we remove this delay?

Reply:
Unfortunately it is there due to a race condition as discussed here: https://consensys.slack.com/archives/C070G0RCX0T/p1718188641805159